guardrails/utils/telemetry_utils.py

```diff
@@ -551,3 +429,650 @@ def trace_llm_call(
+    @wraps(fn)
+    async def trace_async_step_wrapper(*args, **kwargs) -> Iteration:
+        if not settings.disable_tracing:
+            current_otel_context = context.get_current()
+            tracer = get_tracer()
+            tracer = tracer or trace.get_tracer("guardrails-ai", GUARDRAILS_VERSION)
+
+            with tracer.start_as_current_span(
+                name="step",  # type: ignore
+                context=current_otel_context,  # type: ignore
+            ) as step_span:
+                try:
+                    response = await fn(*args, **kwargs)
+                    add_step_attributes(step_span, response, *args, **kwargs)
+                    return response
+                except Exception as e:
+                    step_span.set_status(status=StatusCode.ERROR, description=str(e))
+                    add_step_attributes(step_span, None, *args, **kwargs)
+                    raise e
+
+        else:
+            return await fn(*args, **kwargs)
+
+    return trace_async_step_wrapper
+
+
+async def trace_async_stream_step_generator(
+    fn: Callable[..., AsyncIterable[ValidationOutcome[OT]]], *args, **kwargs
+) -> AsyncIterable[ValidationOutcome[OT]]:
+    current_otel_context = context.get_current()
+    tracer = get_tracer()
+    tracer = tracer or trace.get_tracer("guardrails-ai", GUARDRAILS_VERSION)
+
+    exception = None
+    with tracer.start_as_current_span(
+        name="step",  # type: ignore
+        context=current_otel_context,  # type: ignore
+    ) as step_span:
+        try:
+            gen = fn(*args, **kwargs)
+            next_exists = True
+            while next_exists:
+                try:
+                    res = await anext(gen)  # type: ignore
+                    yield res
+                except StopIteration:
+                    next_exists = False
+        except Exception as e:
+            step_span.set_status(status=StatusCode.ERROR, description=str(e))
+            exception = e
+        finally:
+            call = safe_get(args, 3, kwargs.get("call_log", None))
+            iteration = call.iterations.last if call else None
+            add_step_attributes(step_span, iteration, *args, **kwargs)
+            if exception:
+                raise exception
+
+
+def trace_async_stream_step(
+    fn: Callable[..., AsyncIterable[ValidationOutcome[OT]]],
+):
+    @wraps(fn)
+    async def trace_async_stream_step_wrapper(
+        *args, **kwargs
+    ) -> AsyncIterable[ValidationOutcome[OT]]:
+        if not settings.disable_tracing:
+            return trace_async_stream_step_generator(fn, *args, **kwargs)
+        else:
+            return fn(*args, **kwargs)
+
+    return trace_async_stream_step_wrapper
+
+
+#########################################
+#### END Runner.step Instrumentation ####
+#########################################
+
+
+#########################################
+### START Runner.call Instrumentation ###
```

Comment:
any reason to not break this file up if were subdividing it like this?

Reply:
nope, will refactor these into the `telemetry` directory and out of `utils`